state.go

```diff
@@ -157,92 +235,67 @@ func (s *State) pipe(data, t string, command []string) (string, error) {
 		os.Environ(),
 		"CLAWS_PIPE_TYPE="+t,
 		"CLAWS_SESSION="+strconv.FormatInt(sessionStarted.UnixNano()/1000, 10),
-		"CLAWS_CONNECTION="+strconv.FormatInt(connectionStarted.UnixNano()/1000, 10),
-		"CLAWS_WS_URL="+s.Conn.URL(),
+		"CLAWS_CONNECTION="+strconv.FormatInt(s.ConnectionStarted.UnixNano()/1000, 10),
+		"CLAWS_WS_URL="+s.wsConn.URL(),
 	)
 	// set up stdin
-	stdin := strings.NewReader(data)
+	stdin := bytes.NewReader(data)
 	c.Stdin = stdin
 
 	// run the command
-	res, err := c.Output()
-	return string(res), err
+	return c.Output()
 }
 
-func (s *State) printToOut(f func(io.Writer, ...interface{}) (int, error), str string) {
-	if s.Settings.Timestamp != "" {
-		str = time.Now().Format(s.Settings.Timestamp) + str
+func (s *State) printToOut(
+	str string,
+	bIndent bool,
+	f func(io.Writer, ...interface{}) (int, error),
+) {
```

Comment:
I saw that you changed this to have better formatting for multi-line output. I like that, however I do think that it is useful only in the following circumstances:

- We need to print a multiline string (including a pretty-printed JSON object)
- There is a timestamp.

I think it looks weird in other occasions. For instance, timestamp + singleline message just seems unnecessary:

```
2022-08-24 13:28:25
  {"type": "ping"}
2022-08-24 13:28:25
  {"type":"pong"}
```

It doesn't even help distinguishing from real communication messages and logs from the program, as the colors on the output serve that purpose.

Then as well without a timestamp, we're indenting with no "upper layer" so to speak? So we're just padding with two spaces with no real reason.

I think it might just be worth to remove bIndent altogether and determine whether to indent based on `strings.IndexByte(str, '\n') >= 0 && len(oSet.Timestamp) > 0`

Reply:
of all the things i've tried, i think never indenting the sends (same line as ts), & always indenting the receives (line below ts) is clearest.  the colors are nice, but not universally distinguishable.

![image](https://user-images.githubusercontent.com/8608535/187065367-1ea619f8-df31-4c7b-a3e0-cf072ea55e36.png)

for the sake of "loose coupling", printToOut() should probably be rather dumb, only do what it is explicitly told, & leave the specifics up to its callers.

Reply:
and here's what it looks like with timestamps:

![image](https://user-images.githubusercontent.com/8608535/187065423-23f1c707-a241-4b1d-87ac-6bfcfb633fc7.png)